python/lsst/ts/MTDome/encoding_tools.py

```diff
@@ -57,4 +756,60 @@ def decode(st):
     -------
         A decoded Python representation of the string.
     """
-    return json.loads(st)
+    data = json.loads(st)
+    validate(data)
+    return data
+
+
+def validate(data):
+    """Validates the data against a JSON schema and logs an error in case the
+    validation fails.
+
+    There are eight schemas: one for the commands, one for each of six status
+    command responses and one for all other command responses. This function
+    determines which schema to use based on the keys in the data. Commands are
+    validated as well to ensure that the simulator receives correct commands
+    and this should be done by other clients too.
+
+    Parameters
+    ----------
+    data: `dict`
+        The data to validate. The format of the dict is explained in the
+        `encode` function.
+
+    Raises
+    ------
+    ValidationError:
+        In case the validation fails.
+    ValueError:
+        In case the retrieved data doesn't contain a known key.
+
+    """
+    try:
+        if "command" in data.keys():
```

Comment:
This code can be significantly simplified. There are a few things I would like to change:

1. put your schema into files. Load them in module (not in any function), such as:
```python
command_schema = json.loadstr(open('command_schema.jschema').read())
```

2. this long if/elif can be settled with a hash you will intialise. Keys would be commands, values schemas. So:
```python
encodings = { 'command' : command_schema, .. 'ThCs' : tchs_status_schema }
```
or even better - if you agree to keep schemas in files, just put filename as value and use something like:
```python
encoding = {k: json.load(open(v).read()) for k, v in encoding_filenames.items()}
```

3. trivially, after you have dictionary where keys are commands and values are schemas, do something like:
```python
for k,v in encoding.items():
     if k in data.keys():
          jsonschema.validate(data, v)

if validation_raises_exception:
    raise RuntimeError(f"Validation failed because no known key found in data {data!r}")

Reply:
Thanks for the great suggestion. I put the schemas in files and very much simplified determining which schema to use.